tests/update/tests.py

```diff
@@ -138,3 +139,41 @@ def test_update_respects_to_field(self):
         self.assertEqual(bar_qs[0].foo_id, a_foo.target)
         bar_qs.update(foo=b_foo)
         self.assertEqual(bar_qs[0].foo_id, b_foo.target)
+
+    def test_update_annotated_queryset(self):
+        """
+        Update of a queryset which has been annotated.
+        """
+        # Trivial annotated update first
+        qs = DataPoint.objects.annotate(alias=F('value'))
+        updated = qs.update(another_value='foo')
+        self.assertEqual(updated, 3)
+        # Update where annotation is used for filtering
+        qs = DataPoint.objects.annotate(alias=F('value'))
+        qs = qs.filter(alias='apple')
+        updated = qs.update(another_value='foo')
+        self.assertEqual(updated, 1)
+        # Update where annotation is used in update parameters
+        qs = DataPoint.objects.annotate(alias=F('value'))
+        updated = qs.update(another_value=F('alias'))
+        self.assertEqual(updated, 3)
+
+    def test_update_annotated_multi_table_queryset(self):
+        """
+        Update of a queryset which has been annotated and involves
+        multiple tables. Regression test for #19513 and #18580
+        """
+        # Trivial annotated update first
+        qs = DataPoint.objects.annotate(related_count=Count('relatedpoint'))
+        updated = qs.update(value='Foo')
+        self.assertEqual(updated, 3)
+        # Update where annotation is used for filtering
+        qs = DataPoint.objects.annotate(related_count=Count('relatedpoint'))
+        qs = qs.filter(related_count=1)
+        updated = qs.update(value='Foo')
+        self.assertEqual(updated, 1)
+        # Update where annotation is used in update parameters
+        # #26539 - This isn't forbidden but also doesn't generate proper SQL
+        # qs = RelatedPoint.objects.annotate(data_name=F('data__name'))
+        # updated = qs.update(name=F('data_name'))
```

Comment:
It raises an `OperationalError` because it doesn't fail until the DB tries to execute the invalid SQL. I opened ticket 26539 to cover that, I just left the test in this code because it logically came up when following the pattern of tests (trivial, filtered by annotation, annotation in update params) that all passed for a single table annotation but it failed here for a multi-table annotation.